client/lib/perfmon/index.js

```diff
@@ -0,0 +1,178 @@
+/** 
+ * Internal dependencies
+ */
+import analytics from 'analytics';
+
+/**
+ * External dependencies
+ */
+import each from 'lodash/collection/each';
+import remove from 'lodash/array/remove';
+import debounce from 'lodash/function/debounce';
+import page from 'page';
+
+var debug = require( 'debug' )( 'calypso:perfmon' );
+
+const PLACEHOLDER_CLASSES = [
+	'placeholder',
+	'pulsing-dot is-active',
+	'is-loading'
+];
+
+const PLACEHOLDER_MATCHER = PLACEHOLDER_CLASSES.map(function(clazz) { return `[class*='${clazz}']`; }).join(', ');
+const OBSERVE_ROOT = document.getElementById('wpcom');
+
+let activePlaceholders = [];
+let placeholdersVisibleStart = null;
+
+var perfmon = {
+
+	init: function() {
+		var MutationObserver = window.MutationObserver || window.WebKitMutationObserver;
+
+		if ( MutationObserver ) {
+			this.observeDomChanges( MutationObserver );
+		}
+	},
+
+ 	// add listeners for various DOM events - scrolling, mutation and navigation
+ 	// and use these to trigger checks for visible placeholders (and, in the case of mutations,
+ 	// to record new placeholders and remove nodes that are no longer placeholders)
+	observeDomChanges: function( MutationObserver ) {
+
+		// if anything scrolls, check if any of our placeholder elements are in view,
+		// but not more than a few times a second
+		window.addEventListener('scroll', debounce(this.checkForVisiblePlaceholders.bind( this, 'scroll' ), 200), true);
+
+		// if the user navigates, stop the current event and proceed to the next one
+		page( function( context, next ) {
+			// send "placeholder-wait-navigated" event
+			this.checkForVisiblePlaceholders( 'navigate' );
+			next();
+		}.bind( this ) );
+
+		// this is fired for matching mutations (childList and class attr changes)
+		var observer = new MutationObserver(function(mutations, observer) {
+			
+			// record all the nodes that match our placeholder classes in the "activePlaceholders" array
+			mutations.forEach( this.recordPlaceholders.bind( this ) );
+
+			// remove any nodes from activePlaceholders that are no longer placeholders
+			// check each node for:
+			// a. whether it's still in the DOM at all, and if so:
+			// b. whether it still has a placeholder class
+			var removed = remove( activePlaceholders, function( node ) {
+				return !OBSERVE_ROOT.contains( node ) || !this.isPlaceholder( node );
+			}.bind( this ) );
+
+			this.checkForVisiblePlaceholders( 'mutation' );
+
+		}.bind( this ) );
+
+		observer.observe(OBSERVE_ROOT, {
+		  subtree: true,
+		  attributes: true,
+		  childList: true,
+		  attributeFilter: ['class']
+		});
+	},
+
+	// check if there are any placeholders on the screen,
+	// and trigger a timing event when all the placeholders are gone or the user
+	// has navigated
+	checkForVisiblePlaceholders: function( trigger ) {
+
+		// determine how many placeholders are active in the viewport
+		var visibleCount = 0;
+		activePlaceholders.forEach( function( node ) {
```

Comment:
reduce might be more idiomatic here

so `const visibleCount = activePlaceholders.reduce( function( count, node ) { return count + ( this.isElementVisibleInViewport( node ) ? 1 : 0 ); }.bind(this), 0 )`